R/splitRaster.R

```diff
@@ -80,69 +92,139 @@ setGeneric("splitRaster", function(x, nx, ny, buffer) {
 setMethod(
   "splitRaster",
   signature = signature(x = "RasterLayer", nx = "integer",
-                        ny = "integer", buffer = "numeric"),
-  definition = function(x, nx, ny, buffer) {
+                        ny = "integer", buffer = "numeric",
+                        savePath = "character"),
+  definition = function(x, nx, ny, buffer, savePath) {
     if (length(buffer) > 2) {
       warning("buffer contains more than 2 elements - only the first two will be used.")
       buffer <- buffer[1:2]
     } else if (length(buffer) == 1) {
       buffer <- c(buffer, buffer)
     }
-
     if (buffer[1] < 1) {
       buffer[1] <- ceiling((buffer[1]*(xmax(x) - xmin(x))/nx)/xres(x))
     }
     if (buffer[2] < 1) {
       buffer[2] <- ceiling((buffer[2]*(ymax(x) - ymin(x))/ny)/yres(x))
     }
     ext <- extent(x)
-    tiles <- vector("list", length = nx*ny)
+    extents <- vector("list", length = nx*ny)
+    if(!dir.exists(savePath)) {dir.create(savePath)}
     n <- 1L
     for (i in seq_len(nx) - 1L) {
       for (j in seq_len(ny) - 1L) {
         x0 <- ext@xmin + i*((ext@xmax - ext@xmin) / nx) - buffer[1]*xres(x)
         x1 <- ext@xmin + (i + 1L)*((ext@xmax - ext@xmin) / nx) + buffer[1]*xres(x)
         y0 <- ext@ymin + j*((ext@ymax - ext@ymin) / ny) - buffer[2]*yres(x)
         y1 <- ext@ymin + (j + 1L)*((ext@ymax - ext@ymin) / ny) + buffer[2]*yres(x)
-        tiles[[n]] <- crop(x, extent(x0, x1, y0, y1))
+        extents[[n]] <- extent(x0, x1, y0, y1)
         n <- n + 1L
       }
     }
+    cl <- tryCatch(getCluster(), error = function(x) NULL)
+    on.exit(if (!is.null(cl)) returnCluster())
+    if (!is.null(cl)) {
+      parFun <- "clusterApply"
+      cropNew <- function(ras, x0 = x, ...) {
+        crop(x = ras,y = x0)
+      }
+      args <- list(cl = cl, x = extents,
+                   fun = cropNew, ras = x)
+
+      # args <- list(seq = extents, fun = crop, r)
+      # args1 <- list(cl = cl, y = extents, fun = crop, x = r)
+      # args <- append(list(cl = cl), args)
+    } else {
+      parFun <- "lapply"
+      args <- list(extents, FUN = crop, x = r)
+    }
+    tiles <- do.call(get(parFun), args)
+    for(i in 1:length(tiles)){
```

Comment:
something to chew on...

from the example for `mergeRaster`:

```r
#' m0 <- do.call(merge, y0)
#' all.equal(m0, r) ## should be TRUE but fails
#'
#' m1 <- do.call(merge, y1)
#' all.equal(m1, r) ## should be TRUE but fails
#'
#' m2 <- do.call(merge, y2)
#' all.equal(m2, r) ## TRUE
```

could be related to how the buffer argument is interpreted?